src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,539 @@
+        public MongoAWSAuthenticator(string username, IEnumerable<KeyValuePair<string, string>> properties)
+            : this(username, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+            : base(CreateMechanism(credential, properties, randomByteGenerator, clock))
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, clock))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGODB-AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            ValidateMechanismProperties(properties);
+
+            var awsCredentials =
+                CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties) ??
+                CreateAwsCredentialsFromEnvironmentVariables() ??
+                CreateAwsCredentialsFromEcsResponse() ??
+                CreateAwsCredentialsFromEc2Response();
+
+            if (awsCredentials == null)
+            {
+                throw new InvalidOperationException("A MONGODB-AWS credential must have an access key ID.");
+            }
+
+            var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+
+            return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, clock);
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
```

Comment:
Refactor this method to:

    private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString password, IEnumerable<KeyValuePair<string, string>> properties)
    {
        ValidateMechanismProperties(properties);
        var sessionToken = ExtractSessionTokenFromMechanismProperties(properties);

        if (username == null && password == null && sessionToken == null)
        {
            return null;
        }
        if (password != null && username == null)
        {
            throw new InvalidOperationException("When using MONGODB-AWS authentication if a password is provided via settings then a username must be provided also.");
        }
        if (username != null && password == null)
        {
            throw new InvalidOperationException("When using MONGODB-AWS authentication if a username is provided via settings then a password must be provided also.");
        }
        if (sessionToken != null && (username == null || password == null))
        {
            throw new InvalidOperationException("When using MONGODB-AWS authentication if a session token is provided via settings then a username and password must be provided also.");
        }

        return new AwsCredentials(accessKeyId: username, secretAccessKey: password, sessionToken);
    }

Reply:
Note: the if statements duplicate similar ones in the next method but on purpose because the error messages are more specific to each situation.